ARCs/arc-0020.md

```diff
@@ -0,0 +1,870 @@
+of royalties or [vesting](https://en.wikipedia.org/wiki/Vesting)).
+
+Enforcing transfer conditions require freezing the asset and transferring it
+through a clawback operation -- which results in a process that is opaque to
+users and wallets and a bad experience for the users.
+
+The Smart ASA defined by this ARC extends the ASA to increase its expressiveness
+and its flexibility. Developers can now introduce additional business logic
+around its operations. Wallets, dApps, and SDKs can recognize Smart ASAs and
+adjust their flows and user experiences accordingly.
+
+## Specification
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC
+2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+The following sections describe:
+
+- The ABI interface for a controlling Smart Contract (the Smart Contract that
+  controls a Smart ASA).
+- The metadata required to denote a Smart ASA and define the association between
+  an ASA and its controlling Smart Contract.
+
+### ABI Interface
+
+The ABI interface specified here draws inspiration from the transaction
+[reference](https://developer.algorand.org/docs/get-details/asa/#asset-functions)
+of an Algorand Standard Asset (ASA).
+
+To provide a unified and familiar interface between the Algorand Standard Asset
+and the Smart ASA, method names and parameters have been adapted to the ABI
+types but left otherwise unchanged.
+
+#### Asset Creation
+
+```json
+{
+  "name": "asset_create",
+  "args": [
+    { "type": "uint64", "name": "total" },
+    { "type": "uint32", "name": "decimals" },
+    { "type": "bool", "name": "default_frozen" },
+    { "type": "string", "name": "unit_name" },
+    { "type": "string", "name": "name" },
+    { "type": "string", "name": "url" },
+    { "type": "byte[]", "name": "metadata_hash" },
+    { "type": "address", "name": "manager_addr" },
+    { "type": "address", "name": "reserve_addr" },
+    { "type": "address", "name": "freeze_addr" },
+    { "type": "address", "name": "clawback_addr" }
+  ],
+  "returns": { "type": "uint64" }
+}
+```
+
+Calling `asset_create` creates a new Smart ASA and returns the identifier of the
+ASA. The [metadata section](#metadata) describes its required properties.
+
+> Upon a call to `asset_create`, a reference implementation SHOULD:
+>
+> - Mint an Algorand Standard Asset (ASA) that MUST specify the properties
+>   defined in the [metadata section](#metadata). In addition:
+>   - The `manager`, `reserve` and `freeze` addresses SHOULD be set to the
+>     account of the controlling Smart Contract.
+>   - The remaining fields are left to the implementation, which MAY set `total`
+>     to `2 ** 64 - 1` to enable dynamically increasing the max circulating
+>     supply of the Smart ASA.
+>   - `name` and `unit_name` MAY be set to `SMART-ASA` and `S-ASA`, to
```

Comment:
Most wallets, explorers, etc. are going to display these names so I'm not sure if this is a good recommendation (most end users aren't going to care). Maybe this can be like the ARC-3 recommendation where you add `@arc20` to name or `#arc20` to URL and clients shouldn't display it?

Reply:
I am open to follow this. The trouble is that if you later update the name (smart ASA allows that), what's specified in the ASA will NOT be updated. How do we feel about that?

Reply:
My gut feel is that the name here should be a helpful placeholder.

If you have a UI that asynchronously loads the information it has to hand as it has then it can:

* Show the asset name (from the ASA) first (potentially stripping out `@arc20` if present?)
* Make a call to the underlying smart contract (e.g. by reverse lookup from clawback, or by using the URL field maybe? I note: the reference implementation I found hidden in Beaker used URL field to store a lookup of the app id)
* Once that call came back the UI could cache that value and show the proper name

I think that's a much better user experience for users to see a relevant "holding name" than seeing "SMART-ASA".